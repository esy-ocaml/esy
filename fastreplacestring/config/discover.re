module C = Configurator.V1;

let () =
  C.main(~name="configure fastreplacestring", c => {
    let flags =
      switch (C.ocaml_config_var(c, "system")) {
      | Some("mingw64") => ["-ccopt", "-lstdc++"]
      | Some(_)
      | None => ["-ccopt", "-Ofast", "-ccopt", "-lstdc++"]
      };

    C.Flags.write_sexp("dune.flags", flags);

    let cflags = ["-lstdc++"];
    C.Flags.write_sexp("dune.cflags", cflags);

    let cxx_flags =
      switch (C.ocaml_config_var(c, "system")) {
      /* link statically on windows so we don't have to ship mingw64.dll */
      | Some("mingw64") => ["-static", "-static-libgcc", "-static-libstdc++"]
      | Some("macosx") => ["-lstdc++", "-x", "c++"]
      | Some(_)
      | None => ["-lstdc++"]
      };

    C.Flags.write_sexp("dune.cxx_flags", cxx_flags);
  });
