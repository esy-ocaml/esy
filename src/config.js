/**
 * @flow
 */

import type {Reporter, BuildSpec, Config, BuildPlatform, StoreTree, Store} from './types';
import * as os from 'os';
import invariant from 'invariant';
import {STORE_BUILD_TREE, STORE_INSTALL_TREE, STORE_STAGE_TREE} from './constants';
import * as S from './store';
import * as path from './lib/path';
import * as crypto from './lib/crypto';

const NUM_CPUS = os.cpus().length;

export const ESYB_COMMAND = require.resolve('@esy-ocaml/esyb/esyb');
export const FLOCK_COMMAND = require.resolve('@esy-ocaml/flock/flock');
export const FASTREPLACESTRING_COMMAND = require.resolve(
  'fastreplacestring/.bin/fastreplacestring.exe',
);

function _create<Path: path.Path>(
  sandboxPath: Path,
  store: Store<Path>,
  localStore: Store<Path>,
  importPaths: Array<path.AbsolutePath>,
  buildPlatform,
  reporter: Reporter,
): Config<Path> {
  const genStorePath = (tree: StoreTree, build: BuildSpec, segments: string[]) => {
    const path: Path =
      build.sourceType === 'immutable'
        ? store.getPath(tree, build, ...segments)
        : localStore.getPath(tree, build, ...segments);
    return path;
  };

  function getSourcePath(build: BuildSpec, ...segments): Path {
    const {sourcePath} = build;
    if (sourcePath.charAt(0) === '/') {
      return (path.join(sourcePath, ...segments): any);
    } else {
      return (path.join(sandboxPath, sourcePath, ...segments): any);
    }
  }

  const buildConfig: Config<Path> = {
    reporter,
    sandboxPath,
    store,
    localStore,
    buildPlatform,
    importPaths,
    buildConcurrency: NUM_CPUS,

    getSourcePath,

    getRootPath: (build: BuildSpec, ...segments) => {
      if (build.buildType === 'in-source') {
        return genStorePath(STORE_BUILD_TREE, build, segments);
      } else if (build.buildType === '_build') {
        if (build.sourceType === 'immutable') {
          return genStorePath(STORE_BUILD_TREE, build, segments);
        } else if (build.sourceType === 'transient') {
          return getSourcePath(build, ...segments);
        } else if (build.sourceType === 'root') {
          return getSourcePath(build, ...segments);
        }
      } else if (build.buildType === 'out-of-source') {
        return getSourcePath(build, ...segments);
      }
      invariant(false, 'Impossible happened');
    },
    getBuildPath: (build: BuildSpec, ...segments) =>
      genStorePath(STORE_BUILD_TREE, build, segments),
    getInstallPath: (build: BuildSpec, ...segments) =>
      genStorePath(STORE_STAGE_TREE, build, segments),
    getFinalInstallPath: (build: BuildSpec, ...segments) =>
      genStorePath(STORE_INSTALL_TREE, build, segments),

    prettifyPath: (p: string) => {
      if (p.indexOf(store.path) === 0) {
        const relative = p.slice(store.path.length);
        return path.join(store.prettyPath, relative);
      } else {
        return p;
      }
    },

    getSandboxPath: (requests: Array<string>) => {
      // TODO: normalize requests?
      requests = requests.slice(0);
      requests.sort();
      const id = crypto.hash(requests.join(' '));
      // TODO: how to get prefix? is it availble in any config?
      const prefixPath = path.dirname(store.path);
      return path.join(prefixPath, 'sandbox', id);
    },
  };
  return buildConfig;
}

export function create(params: {
  reporter: Reporter,
  storePath: string,
  sandboxPath: string,
  buildPlatform: BuildPlatform,
  importPaths?: Array<string>,
}): Config<path.AbstractPath> {
  const {reporter, storePath, sandboxPath, buildPlatform, importPaths = []} = params;
  const store = S.forAbstractPath(storePath);
  const localStore = S.forAbstractPath(
    path.join(sandboxPath, 'node_modules', '.cache', '_esy', 'store'),
  );
  return _create(
    path.abstract(sandboxPath),
    store,
    localStore,
    importPaths.map(path.absolute),
    buildPlatform,
    reporter,
  );
}

export function createForPrefix(params: {
  reporter: Reporter,
  prefixPath: string,
  sandboxPath: string,
  buildPlatform: BuildPlatform,
  importPaths?: Array<string>,
}): Config<path.AbsolutePath> {
  const {reporter, prefixPath, sandboxPath, buildPlatform, importPaths = []} = params;
  const store = S.forPrefixPath(prefixPath);
  const localStore = S.forAbsolutePath(
    path.join(sandboxPath, 'node_modules', '.cache', '_esy', 'store'),
  );
  return _create(
    path.absolute(sandboxPath),
    store,
    localStore,
    importPaths.map(path.absolute),
    buildPlatform,
    reporter,
  );
}
