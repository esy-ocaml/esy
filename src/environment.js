/**
 * @flow
 */

import type {Environment, EnvironmentBinding} from './types';

import * as shell from './lib/shell.js';
import * as os from 'os';

// X platform newline
const EOL = os.EOL;

/**
 * Print environment as a bash script.
 */
export function printEnvironment(env: Environment) {
  const groupsByBuild = new Map();

  const groups = [];

  for (const item of env) {
    const header =
      item.origin != null
        ? `${item.origin.name}@${item.origin.version} ${item.origin.packagePath}`
        : 'Esy Sandbox';
    const curGroup = groups[groups.length - 1];
    if (groups.length === 0 || curGroup.header !== header) {
      const curGroup = {header, env: [item]};
      groups.push(curGroup);
    } else {
      curGroup.env.push(item);
    }
  }

  return Array.from(groups)
    .map(group => {
      const headerLines = [`# ${group.header}`];
      // TODO: add error rendering here
      // const errorLines = group.errors.map(err => {
      //   return '# [ERROR] ' + err;
      // });
      const envVarLines = group.env.map(item => {
        // TODO: escape " in values
        const exportLine = `export ${item.name}=${shell.doubleQuote(item.value)}`;
        return exportLine;
      });
      return headerLines.concat(envVarLines).join(EOL);
    })
    .join(EOL);
}

/**
 * Eval environment into a mapping of key-values.
 */
export function evalEnvironment(env: Environment): Map<string, string> {
  const envMap = new Map();
  env.forEach(item => {
    const {value} = shell.expand(item.value, name => {
      const item = envMap.get(name);
      return item != null ? item : undefined;
    });
    if (value != null) {
      envMap.set(item.name, value);
    }
  });
  return envMap;
}
