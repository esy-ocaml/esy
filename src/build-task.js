/**
 * @flow
 */

import type {
  BuildSandbox,
  BuildSpec,
  Config,
  BuildTask,
  BuildEnvironment,
  BuildPlatform,
} from './types';

import {substituteVariables} from 'var-expansion';

import {doubleQuote} from './lib/shell';
import * as path from './lib/path';
import {normalizePackageName, mergeIntoMap, mapValuesMap} from './util';
import * as Graph from './graph';
import * as Env from './environment';

type BuildTaskParams = {
  env?: BuildEnvironment,
};

/**
 * Produce a task graph from a build spec graph.
 */
export function fromBuildSpec(
  rootBuild: BuildSpec,
  config: Config<path.Path>,
  params?: BuildTaskParams = {},
): BuildTask {
  const {task} = Graph.topologicalFold(
    rootBuild,
    (dependencies, allDependencies, spec) => {
      const scopes = computeScopes(dependencies, allDependencies, spec);
      const task = createTask(scopes);
      return {spec, scopes, task};
    },
  );

  function computeScopes(dependencies, allDependencies, spec) {
    // scope which is used to eval exported variables
    const evalScope = getEvalScope(spec, dependencies, config);
    // global env vars exported from a spec
    const globalScope = new Map();
    // local env vars exported from a spec
    const localScope = new Map();
    for (const name in spec.exportedEnv) {
      const envConfig = spec.exportedEnv[name];
      const value = renderWithScope(envConfig.val, evalScope).rendered;
      const item = {
        name,
        value,
        spec,
        builtIn: false,
        exported: true,
        exclusive: Boolean(envConfig.exclusive),
      };
      if (envConfig.scope === 'global') {
        globalScope.set(name, item);
      } else {
        localScope.set(name, item);
      }
    }
    const scopes = {
      spec,
      localScope,
      globalScope,
      dependencies,
      allDependencies,
    };
    return scopes;
  }

  function createTask(scopes): BuildTask {
    const env = new Map();
    const ocamlfindDest = config.getInstallPath(scopes.spec, 'lib');

    const OCAMLPATH = [];
    const PATH = [];
    const MAN_PATH = [];

    for (const dep of scopes.allDependencies.values()) {
      OCAMLPATH.push(config.getFinalInstallPath(dep.spec, 'lib'));
      PATH.push(config.getFinalInstallPath(dep.spec, 'bin'));
      MAN_PATH.push(config.getFinalInstallPath(dep.spec, 'man'));
    }

    // In ideal world we wouldn't need it as the whole toolchain should be
    // sandboxed. This isn't the case unfortunately.
    PATH.push('$PATH');
    MAN_PATH.push('$MAN_PATH');

    evalIntoEnv(env, [
      {
        name: 'OCAMLPATH',
        value: OCAMLPATH.join(getPathsDelimiter('OCAMLPATH', config.buildPlatform)),
        exported: true,
        exclusive: true,
      },
      {
        name: 'OCAMLFIND_DESTDIR',
        value: ocamlfindDest,
        exported: true,
        exclusive: true,
      },
      {
        name: 'OCAMLFIND_LDCONF',
        value: 'ignore',
        exported: true,
        exclusive: true,
      },
      {
        name: 'OCAMLFIND_COMMANDS',
        // eslint-disable-next-line max-len
        value:
          'ocamlc=ocamlc.opt ocamldep=ocamldep.opt ocamldoc=ocamldoc.opt ocamllex=ocamllex.opt ocamlopt=ocamlopt.opt',
        exported: true,
        exclusive: true,
      },
      {
        name: 'PATH',
        value: PATH.join(getPathsDelimiter('PATH', config.buildPlatform)),
        exported: true,
      },
      {
        name: 'MAN_PATH',
        value: MAN_PATH.join(getPathsDelimiter('MAN_PATH', config.buildPlatform)),
        exported: true,
      },
    ]);

    const errors = [];

    // $cur__name, $cur__version and so on...
    mergeIntoMap(env, getBuiltInScope(scopes.spec, config, true));

    // direct deps' local scopes
    for (const dep of scopes.dependencies.values()) {
      mergeIntoMap(env, dep.scopes.localScope);
    }
    // build's own local scope
    mergeIntoMap(env, scopes.localScope);
    // all deps' global scopes merged
    mergeIntoMap(
      env,
      Env.merge(
        Array.from(scopes.allDependencies.values())
          .map(dep => dep.scopes.globalScope)
          .concat(scopes.globalScope),
        evalIntoEnv,
      ),
    );

    if (params != null && params.env != null) {
      evalIntoEnv(env, Array.from(params.env.values()));
    }

    const scope = new Map();
    mergeIntoMap(scope, getEvalScope(scopes.spec, scopes.dependencies, config));
    mergeIntoMap(scope, env);

    const buildCommand = scopes.spec.buildCommand.map(command =>
      renderCommand(command, scope),
    );
    const installCommand = scopes.spec.installCommand.map(command =>
      renderCommand(command, scope),
    );

    return {
      id: scopes.spec.id,
      spec: scopes.spec,
      buildCommand,
      installCommand,
      env,
      scope,
      dependencies: mapValuesMap(scopes.dependencies, dep => dep.task),
      errors,
    };
  }

  function renderCommand(command: Array<string> | string, scope) {
    if (Array.isArray(command)) {
      return {
        command: command.join(' '),
        renderedCommand: command
          .map(command => quoteArgIfNeeded(expandWithScope(command, scope).rendered))
          .join(' '),
      };
    } else {
      return {
        command,
        renderedCommand: expandWithScope(command, scope).rendered,
      };
    }
  }

  return task;
}

function builtInEntry({
  name,
  value,
  spec,
  exclusive = true,
  exported = false,
}: {
  name: string,
  value: string,
  spec?: BuildSpec,
  exclusive?: boolean,
  exported?: boolean,
}) {
  return [name, {name, value, spec, builtIn: true, exclusive, exported}];
}

function builtInEntries(...values) {
  return new Map(values.map(builtInEntry));
}

function getBuiltInScope(
  spec: BuildSpec,
  config: Config<path.Path>,
  currentlyBuilding?: boolean,
): BuildEnvironment {
  const prefix = currentlyBuilding ? 'cur' : normalizePackageName(spec.name);
  const getInstallPath = currentlyBuilding
    ? config.getInstallPath
    : config.getFinalInstallPath;
  return builtInEntries(
    {
      name: `${prefix}__name`,
      value: spec.name,
      spec,
    },
    {
      name: `${prefix}__version`,
      value: spec.version,
      spec,
    },
    {
      name: `${prefix}__root`,
      value:
        currentlyBuilding && config.requiresRootRelocation(spec)
          ? config.getBuildPath(spec)
          : config.getRootPath(spec),
      spec,
    },
    {
      name: `${prefix}__depends`,
      value: Array.from(spec.dependencies.values(), dep => dep.name).join(' '),
      spec,
    },
    {
      name: `${prefix}__target_dir`,
      value: config.getBuildPath(spec),
      spec,
    },
    {
      name: `${prefix}__install`,
      value: getInstallPath(spec),
      spec,
    },
    {
      name: `${prefix}__bin`,
      value: getInstallPath(spec, 'bin'),
      spec,
    },
    {
      name: `${prefix}__sbin`,
      value: getInstallPath(spec, 'sbin'),
      spec,
    },
    {
      name: `${prefix}__lib`,
      value: getInstallPath(spec, 'lib'),
      spec,
    },
    {
      name: `${prefix}__man`,
      value: getInstallPath(spec, 'man'),
      spec,
    },
    {
      name: `${prefix}__doc`,
      value: getInstallPath(spec, 'doc'),
      spec,
    },
    {
      name: `${prefix}__stublibs`,
      value: getInstallPath(spec, 'stublibs'),
      spec,
    },
    {
      name: `${prefix}__toplevel`,
      value: getInstallPath(spec, 'toplevel'),
      spec,
    },
    {
      name: `${prefix}__share`,
      value: getInstallPath(spec, 'share'),
      spec,
    },
    {
      name: `${prefix}__etc`,
      value: getInstallPath(spec, 'etc'),
      spec,
    },
  );
}

function evalIntoEnv<V: {name: string, value: string}>(
  scope: BuildEnvironment,
  items: Array<V>,
) {
  const update = new Map();
  for (const item of items) {
    const nextItem = {
      exported: true,
      exclusive: false,
      builtIn: false,
      ...item,
      value: renderWithScope(item.value, scope).rendered,
    };
    update.set(item.name, nextItem);
  }
  mergeIntoMap(scope, update);
  return scope;
}

function getEvalScope(spec: BuildSpec, dependencies, config): BuildEnvironment {
  const evalScope = new Map();
  for (const dep of dependencies.values()) {
    mergeIntoMap(evalScope, getBuiltInScope(dep.spec, config));
    mergeIntoMap(evalScope, dep.scopes.localScope);
  }
  mergeIntoMap(evalScope, getBuiltInScope(spec, config));
  return evalScope;
}

const FIND_VAR_RE = /\$([a-zA-Z0-9_]+)/g;

export function renderWithScope<T: {value: string}>(
  value: string,
  scope: Map<string, T>,
): {rendered: string} {
  const rendered = value.replace(FIND_VAR_RE, (_, name) => {
    const value = scope.get(name);
    if (value == null) {
      return `\$${name}`;
    } else {
      return value.value;
    }
  });
  return {rendered};
}

export function quoteArgIfNeeded(arg: string): string {
  if (arg.indexOf(' ') === -1 && arg.indexOf("'") === -1 && arg.indexOf('"') === -1) {
    return arg;
  } else {
    return doubleQuote(arg);
  }
}

export function expandWithScope<T: {value: string}>(
  value: string,
  scope: Map<string, T>,
): {rendered: string} {
  const {value: rendered} = substituteVariables(value, {
    env: name => {
      const item = scope.get(name);
      return item != null ? item.value : undefined;
    },
  });
  return {rendered: rendered != null ? rendered : value};
}

export function fromBuildSandbox<Path: path.Path>(
  sandbox: BuildSandbox,
  config: Config<Path>,
  params?: BuildTaskParams,
): BuildTask {
  const env = new Map();
  if (sandbox.env) {
    mergeIntoMap(env, sandbox.env);
  }
  if (params != null && params.env != null) {
    mergeIntoMap(env, params.env);
  }
  return fromBuildSpec(sandbox.root, config, {...params, env});
}

/**
 * Logic to determine how file paths inside of env vars should be delimited.
 * For example, what separates file paths in the `PATH` env variable, or
 * `OCAMLPATH` variable? In an ideal world, the logic would be very simple:
 * `linux`/`darwin`/`cygwin` always uses `:`, and Windows/MinGW always uses
 * `;`, however there's some unfortunate edge cases to deal with - `esy` can
 * take care of all of that for you.
 */
function getPathsDelimiter(envVarName: string, buildPlatform: BuildPlatform) {
  // Error as a courtesy. This means something went wrong in the esy code, not
  // consumer code. Should be fixed ASAP.
  if (envVarName === '' || envVarName.charAt(0) === '$') {
    throw new Error('Invalidly formed environment variable:' + envVarName);
  }
  if (buildPlatform === null || buildPlatform === undefined) {
    throw new Error('Build platform not specified');
  }
  // Comprehensive pattern matching would be nice to have here!
  return envVarName === 'OCAMLPATH' && buildPlatform === 'cygwin'
    ? ';'
    : buildPlatform === 'cygwin' ||
      buildPlatform === 'linux' ||
      buildPlatform === 'darwin'
      ? ':'
      : ';';
}
