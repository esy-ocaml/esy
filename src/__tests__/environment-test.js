/**
 * @flow
 */

import type {BuildSpec} from '../types';
import {create as createConfig} from '../config';
import {fromBuildSpec} from '../build-task';
import {printEnvironment} from '../environment';

function build({name, exportedEnv, dependencies: dependenciesArray}): BuildSpec {
  const dependencies = new Map();
  for (const item of dependenciesArray) {
    dependencies.set(item.id, item);
  }
  return {
    id: name,
    name,
    version: '0.1.0',
    dependencies,
    exportedEnv,
    mutatesSourcePath: false,
    sourcePath: name,
    sourceType: 'immutable',
    shouldBePersisted: true,
    command: null,
    errors: [],
  };
}

const config = createConfig({
  sandboxPath: '<sandboxPath>',
  storePath: '<storePath>',
  buildPlatform: 'linux',
  readOnlyStorePath: [],
});

const ocaml = build({
  name: 'ocaml',
  exportedEnv: {
    CAML_LD_LIBRARY_PATH: {
      val: '$ocaml__lib/ocaml',
      scope: 'global',
    },
  },
  dependencies: [],
});

const ocamlfind = build({
  name: 'ocamlfind',
  exportedEnv: {
    CAML_LD_LIBRARY_PATH: {
      val: '$ocamlfind__lib/ocaml:$CAML_LD_LIBRARY_PATH',
      scope: 'global',
    },
  },
  dependencies: [ocaml],
});

const lwt = build({
  name: 'lwt',
  exportedEnv: {
    CAML_LD_LIBRARY_PATH: {
      val: '$lwt__lib/ocaml:$CAML_LD_LIBRARY_PATH',
      scope: 'global',
    },
  },
  dependencies: [ocaml],
});

describe('printEnvironment()', function() {
  test('printing environment', function() {
    const app = build({
      name: 'app',
      exportedEnv: {
        CAML_LD_LIBRARY_PATH: {
          val: '$app__lib:$CAML_LD_LIBRARY_PATH',
          scope: 'global',
        },
      },
      dependencies: [ocamlfind, lwt],
    });
    const {env} = fromBuildSpec(app, config);
    expect(printEnvironment(env)).toMatchSnapshot();
  });
});
