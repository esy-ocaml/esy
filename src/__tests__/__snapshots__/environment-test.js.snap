// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`printEnvironment() printing environment 1`] = `
"# Esy Sandbox
export OCAMLPATH=\\"<storePath>/i/lwt/lib:<storePath>/i/ocamlfind/lib:<storePath>/i/ocaml/lib\\"
export OCAMLFIND_DESTDIR=\\"<storePath>/s/app/lib\\"
export OCAMLFIND_LDCONF=\\"ignore\\"
export OCAMLFIND_COMMANDS=\\"ocamlc=ocamlc.opt ocamldep=ocamldep.opt ocamldoc=ocamldoc.opt ocamllex=ocamllex.opt ocamlopt=ocamlopt.opt\\"
export PATH=\\"<storePath>/i/lwt/bin:<storePath>/i/ocamlfind/bin:<storePath>/i/ocaml/bin:$PATH\\"
export MAN_PATH=\\"<storePath>/i/lwt/man:<storePath>/i/ocamlfind/man:<storePath>/i/ocaml/man:$MAN_PATH\\"
# app@0.1.0 app
export cur__name=\\"app\\"
export cur__version=\\"0.1.0\\"
export cur__root=\\"<sandboxPath>/app\\"
export cur__depends=\\"ocamlfind lwt\\"
export cur__target_dir=\\"<storePath>/b/app\\"
export cur__install=\\"<storePath>/s/app\\"
export cur__bin=\\"<storePath>/s/app/bin\\"
export cur__sbin=\\"<storePath>/s/app/sbin\\"
export cur__lib=\\"<storePath>/s/app/lib\\"
export cur__man=\\"<storePath>/s/app/man\\"
export cur__doc=\\"<storePath>/s/app/doc\\"
export cur__stublibs=\\"<storePath>/s/app/stublibs\\"
export cur__toplevel=\\"<storePath>/s/app/toplevel\\"
export cur__share=\\"<storePath>/s/app/share\\"
export cur__etc=\\"<storePath>/s/app/etc\\"
export CAML_LD_LIBRARY_PATH=\\"<storePath>/i/app/lib:<storePath>/i/lwt/lib/ocaml:<storePath>/i/ocamlfind/lib/ocaml:<storePath>/i/ocaml/lib/ocaml\\""
`;
